Geometry_Engine/Query/Centroid.cs

```diff
@@ -0,0 +1,247 @@
+                    xc0 -= ((pA.X + pB.X + pC.X) / 3) * triangleArea;
+                    yc0 -= ((pA.Y + pB.Y + pC.Y) / 3) * triangleArea;
+                    zc0 -= ((pA.Z + pB.Z + pC.Z) / 3) * triangleArea;
+                }
+            }
+
+
+            if (!dir)
+            {
+                xc0 = -xc0;
+                yc0 = -yc0;
+                zc0 = -zc0;
+            }
+
+            double curveArea = curve.Area();
+
+            xc = xc0 / curveArea;
+            yc = yc0 / curveArea;
+            zc = zc0 / curveArea;
+
+            return new Point { X = xc, Y = yc, Z = zc };
+
+        }
+
+
+        /***************************************************/
+
+        public static Point Centroid(this PolyCurve curve)
+        {
+            if (!curve.IsPlanar())
+            {
+                Reflection.Compute.RecordError("Input must be planar.");
+                return null;
+            }
+            else if (!curve.IsClosed())
+            {
+                Reflection.Compute.RecordError("Curve is not closed.");
+                return null;
+            }
+            else if (curve.IsSelfIntersecting())
+            {
+                Reflection.Compute.RecordWarning("Curve is self intersecting");
+                return null;
+            }
+            
+           List<Point> pts = new List<Point> { curve.Curves[0].IStartPoint() };
+            foreach (ICurve crv in curve.SubParts())
+            {
+                if (crv is Line)
+                    pts.Add((crv as Line).End);
+                else if (crv is Arc)
+                    pts.Add(crv.IEndPoint());
+                else
+                    throw new NotImplementedException();
+            }
+
+            double xc, yc, zc;
+            double xc0 = 0, yc0 = 0, zc0 = 0;
+
+            Vector normal = Normal(curve);
+
+            Point pA = pts[0];
+            Point pB0 = pts[1];
+            Point pC0 = pts[2];
+
+            Vector firstNormal;
+
+            firstNormal = CrossProduct(pB0 - pA, pC0 - pA);
+
+            Boolean dir = DotProduct(normal, firstNormal) > 0;
+
+            for (int i = 1; i < pts.Count - 2; i++)
+            {
+
+                Point pB = pts[i];
+                Point pC = pts[i + 1];
+
+                double triangleArea = Area(pB - pA, pC - pA);
+
+                if (DotProduct(CrossProduct(pB - pA, pC - pA), firstNormal) > 0)
+                {
+                    xc0 += ((pA.X + pB.X + pC.X) / 3) * triangleArea;
+                    yc0 += ((pA.Y + pB.Y + pC.Y) / 3) * triangleArea;
+                    zc0 += ((pA.Z + pB.Z + pC.Z) / 3) * triangleArea;
+                }
+                else
+                {
+                    xc0 -= ((pA.X + pB.X + pC.X) / 3) * triangleArea;
+                    yc0 -= ((pA.Y + pB.Y + pC.Y) / 3) * triangleArea;
+                    zc0 -= ((pA.Z + pB.Z + pC.Z) / 3) * triangleArea;
+                }
+            }
+
+            foreach (ICurve crv in curve.Curves)
```

Comment:
Need to call `curve.SubParts()` here, same as for the Normal to get out arc and line segments, as well as any potential polycurves inside the polycurves.  Should not be a goto solution ever, but there is nothing in the class stopping it.

Could also store all the subparts as used in the forloop above on line 134 to avoid having to do it twice.